File: backend/app/api/serviceability.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from app.engine.serviceability_checks import ServiceabilityChecks

router = APIRouter()

class DeflectionCheckRequest(BaseModel):
    span: float
    actual_deflection: float
    member_type: str = "beam"
    support_condition: str = "simply_supported"
    loading_type: str = "live"

class CrackWidthCheckRequest(BaseModel):
    stress_steel: float
    cover: float
    bar_diameter: float
    spacing: float
    exposure_condition: str = "moderate"

class VibrationCheckRequest(BaseModel):
    natural_frequency: float
    floor_type: str = "office"
    damping_ratio: float = 0.05

class PunchingShearCheckRequest(BaseModel):
    column_size: float
    slab_thickness: float
    column_load: float
    fck: Optional[float] = None

class FatigueCheckRequest(BaseModel):
    stress_range: float
    n_cycles: float
    material: str = "steel"
    detail_category: str = "C"

class SlendernessCheckRequest(BaseModel):
    length: float
    radius_of_gyration: float
    member_type: str = "column"
    end_conditions: str = "pinned_pinned"

@router.post("/deflection")
def check_deflection(request: DeflectionCheckRequest):
    """Check deflection limits"""
    try:
        checker = ServiceabilityChecks()
        result = checker.deflection_check(
            request.span,
            request.actual_deflection,
            request.member_type,
            request.support_condition,
            request.loading_type
        )
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/crack-width")
def check_crack_width(request: CrackWidthCheckRequest):
    """Check crack width limits"""
    try:
        checker = ServiceabilityChecks()
        result = checker.crack_width_check(
            request.stress_steel,
            request.cover,
            request.bar_diameter,
            request.spacing,
            request.exposure_condition
        )
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/vibration")
def check_vibration(request: VibrationCheckRequest):
    """Check vibration limits"""
    try:
        checker = ServiceabilityChecks()
        result = checker.vibration_check(
            request.natural_frequency,
            request.floor_type,
            request.damping_ratio
        )
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/punching-shear")
def check_punching_shear(request: PunchingShearCheckRequest):
    """Check punching shear"""
    try:
        checker = ServiceabilityChecks()
        result = checker.punching_shear_check(
            request.column_size,
            request.slab_thickness,
            request.column_load,
            request.fck
        )
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/fatigue")
def check_fatigue(request: FatigueCheckRequest):
    """Check fatigue"""
    try:
        checker = ServiceabilityChecks()
        result = checker.fatigue_check(
            request.stress_range,
            request.n_cycles,
            request.material,
            request.detail_category
        )
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/slenderness")
def check_slenderness(request: SlendernessCheckRequest):
    """Check slenderness ratio"""
    try:
        checker = ServiceabilityChecks()
        result = checker.slenderness_check(
            request.length,
            request.radius_of_gyration,
            request.member_type,
            request.end_conditions
        )
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/src/lib/api/serviceability.ts
import { apiClient } from './client';
import type { ApiResponse } from '@/types/api';

export interface DeflectionCheckRequest {
  span: number;
  deflection: number;
  load_type: 'dead' | 'live' | 'total';
  member_type: 'beam' | 'slab' | 'cantilever';
}

export interface CrackWidthCheckRequest {
  stress: number;
  cover: number;
  bar_diameter: number;
  spacing: number;
  exposure: 'mild' | 'moderate' | 'severe';
}

export interface VibrationCheckRequest {
  natural_frequency: number;
  floor_type: 'office' | 'residential' | 'hospital';
  damping_ratio?: number;
}

export interface PunchingShearCheckRequest {
  slab_thickness: number;
  column_dimensions: Record<string, number>;
  punching_force: number;
  concrete_grade: number;
}

export interface FatigueCheckRequest {
  stress_range: number;
  cycles: number;
  detail_category: string;
}

export interface SlendernessCheckRequest {
  length: number;
  radius_of_gyration: number;
  end_conditions: string;
}

export const serviceabilityApi = {
  checkDeflection: async (data: DeflectionCheckRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/serviceability/deflection', data);
    return response.data.data!;
  },

  checkCrackWidth: async (data: CrackWidthCheckRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/serviceability/crack-width', data);
    return response.data.data!;
  },

  checkVibration: async (data: VibrationCheckRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/serviceability/vibration', data);
    return response.data.data!;
  },

  checkPunchingShear: async (data: PunchingShearCheckRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/serviceability/punching-shear', data);
    return response.data.data!;
  },

  checkFatigue: async (data: FatigueCheckRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/serviceability/fatigue', data);
    return response.data.data!;
  },

  checkSlenderness: async (data: SlendernessCheckRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/serviceability/slenderness', data);
    return response.data.data!;
  },
};
